### PR TITLE
enhance logger rotator

### PR DESCRIPTION
**Pull request by nejisama**
1. get time zone every rotate time
2. check file exists before rotate, if exists, add a generation suffix
3. recalcuate rotate interval when default global roller updated by config
4. support recover  logger registered if needed.

### CODE REVIEW

File: log/logger.go [hunk 1]
```diff
@@ -368,6 +371,12 @@ func (l *Logger) Fatalln(args ...interface{}) {
 	os.Exit(1)
 }
 
+func (l *Logger) calcInterval(now time.Time) time.Duration {
+	// caculate the next time need to rotate
```
Comment (taoyuanyuan) [inline log/logger.go:375]: calculate

File: log/logger.go [hunk 2]
```diff
@@ -405,11 +423,12 @@ func doRotateFunc(l *Logger, interval time.Duration) {
 			go l.Reopen()
 
 			if interval == 0 { // recaculate interval
-				interval = time.Duration(l.roller.MaxTime-(now.Unix()+int64(localOffset))%l.roller.MaxTime) * time.Second
+				interval = l.calcInterval(now)
```
Comment (taoyuanyuan) [inline log/logger.go:426]: // recalculate

File: log/roller.go [hunk 3]
```diff
@@ -125,12 +141,30 @@ func DefaultRoller() *Roller {
...
-		os.Rename(l.FileName, l.FileName+"."+l.CreateTime.Format("2006-01-02"))
+		filename = l.FileName + "." + l.CreateTime.Format("2006-01-02")
 	} else {
-		os.Rename(l.FileName, l.FileName+"."+l.CreateTime.Format("2006-01-02_15"))
+		filename = l.FileName + "." + l.CreateTime.Format("2006-01-02_15")
+	}
+
+	name := filename
+	// if rollerFile exists, add a generational name
+	for {
```
Comment (taoyuanyuan) [inline log/roller.go:155]: 设置个次数吧 兜个底

File: utils/goroutine.go [hunk 4]
```diff
@@ -19,28 +19,32 @@ package utils
...
 // GoWithRecover wraps a `go func()` with recover()
 func GoWithRecover(handler func(), recoverHandler func(r interface{})) {
 	go func() {
 		defer func() {
 			if r := recover(); r != nil {
-				// TODO: log
-				if !debugIgnoreStdout {
-					fmt.Fprintf(os.Stderr, "%s goroutine panic: %v\n%s\n", CacheTime(), r, string(debug.Stack()))
-				}
+				recoverLogger(os.Stderr, r)
```
Comment (taoyuanyuan) [inline utils/goroutine.go:42]: 需要包装一个 recover，不然recoverLogger panic了，mosn就跪了

Comment (nejisama) [reply utils/goroutine.go:45]: 这里本来就是捕捉panic了，这里再recover嵌套有点多了。 就先这样吧。加了一个注释说明一下

**Review by taoyuanyuan — APPROVED** [review]
LGTM

**Pull request merged**